### PR TITLE
Feature/complete align and fine tune

### PR DESCRIPTION
**Pull request by richard0000**
# What was done?
### 6. Save the changes on alignments to DB

- Iterate thorugh all the alignments to save the changes to DB.
- Make the button to save available only if not loading and if
there are any changes performed.
- Save the mappedTerms as a many to many relation

### 7. Add synthetic element on UI & Save synthetic element added to the spine

This process consts of saving both:
- The spine term, which is a property in the specification marked
as spine for the current domain, and
- The alignment (mappingTerm) which specifies which mappedTerm/s
from the uploaded specification is going to be aligned with the
new synthetic term and with which predicate.
- Make the button to save available only if not loading and if there are any changes performed.

### 12. More

- Filter by "changed" alignments before save, so we don't
save all the alignments every time.
- Add transactions to handle alignment updates.
- Improve error handling.
- Remove memory leak issues on Collapsible and ExpandableOptions components
- Simplify the update mapped terms logic
- Replace "Animated" component: The Animated component is causing problems on
"ComponentWillReceiveProps". That method is deprecated in React. Use instead,
"Styled Components" with "React Animations". Create a component to wrap all
animations.
- Propagate the name dropped on the left (property name) on Synthetic element
added to the spine.
- Permit only 1 term to drag when adding synthetic

# Screenshots
align-n-fine tune1

### CODE REVIEW

File: app/models/mapping_term.rb [hunk 1]
```diff
@@ -14,4 +14,21 @@ class MappingTerm < ApplicationRecord
...
+  # @description: Give the params the correct format. This is in case we receive an array of
+  #   mapped_terms. We need ot be instantiate each term in order to persist
+  # @param [Array] mapped_term_ids: A collection of ids representing the terms that are
+  #   being mapped in this alignment
+  ###
+  def update_mapped_terms mapped_term_ids
+    # Instantiate each mapped term
+    mapped_terms_collection =
+      (mapped_term_ids - mapped_terms.pluck(:id)).map do |mapped_term_id|
+        Term.find(mapped_term_id)
```
Comment (excelsior) [inline app/models/mapping_term.rb:28]: What's the point of instantiating the terms here? Can't we add them by the IDs:

```ruby
  mapped_term_ids += (mapped_terms.pluck(:id) - mapped_term_ids)
```

File: config/routes.rb [hunk 2]
```diff
@@ -14,12 +14,14 @@
...
       resources :domains, only: [:index, :show]
       resources :mappings, only: [:create, :show, :index, :update]
+      resources :mapping_terms, only: [:update]
       resources :organizations, only: [:index, :show, :create, :update, :destroy]
       resources :predicates, only: [:index]
       resources :roles, only: [:index]
       resources :specifications, only: [:create, :show]
       resources :terms, only: [:show, :update, :destroy]
       resources :vocabularies, only: [:index, :create]
+      resources :mapping_terms, only: [:update]
```
Comment (excelsior) [inline config/routes.rb:24]: Duplicate of line 17.

File: app/controllers/api/v1/mapping_terms_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  ###
+  # @description: Execute the authorization policy
+  ###
+  def authorize_with_policy
+    authorize mapping_term
+  end
+
+  ###
+  # @description: Get the current model record
+  # @return [ActiveRecord]
```
Comment (excelsior) [inline app/controllers/api/v1/mapping_terms_controller.rb:27]: Totally not a biggie, but the returned type is `MappingTerm`, `ActiveRecord` is a module.

**Review by excelsior — CHANGES_REQUESTED** [review]
Thanks, @richard0000 

Works fine, but there's an issue with displaying errors. Also, I've suggested a couple of smallish improvements, hopefully you find them useful.

File: app/models/mapping_term.rb (hunk 1 above)
Comment (richard0000) [reply app/models/mapping_term.rb:28]: @excelsior there's no such attribute `mapped_ter_ids` its the name of the param. We're modeling a "many to many" relationship here, so a valid way to perform the update is by instantiating the Term (or collection of Terms) and then adding to the relationship name (`mapped_terms` in this case).

The other way is by adding a class for the join table, which in this case I consider unnecessary due to the fact that we don't have any additional information to store on such model.

Comment (excelsior) [reply app/models/mapping_term.rb]: Sorry, my comment was somewhat confusing. 

Adding the `has_and_belongs_to_many :mapped_terms` association generates the `mapped_term_ids` and `mapped_term_ids=` methods (among others):
https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_and_belongs_to_many

Hence, this method can be simplified to something like this:

```ruby
def update_mapped_terms(term_ids)
  self.mapped_term_ids += (term_ids - mapped_term_ids)
end
```

File: app/javascript/components/shared/ExpandableOptions.jsx [hunk 4]
```diff
@@ -130,11 +129,19 @@ class ExpandableOptions extends Component {
               >
                 <p>
                   <strong>{selectedOption}</strong>
-                  <span className="cursor-pointer" onClick={() => this.handleExpand()}>
-                    <i className={(this.props.iconClassName || "fas fa-angle-down") + " float-right"}></i>
+                  <span
+                    className="cursor-pointer"
+                    onClick={() => this.handleExpand()}
```
Comment (excelsior) [inline app/javascript/components/shared/ExpandableOptions.jsx:134]: Can't this be simplified to just: `onClick={this.handleExpand}`

File: app/javascript/components/align-and-fine-tune/AlignAndFineTune.jsx [hunk 5]
```diff
@@ -479,26 +479,32 @@ const AlignAndFineTune = (props) => {
...
       await Promise.all(
-        mappingTerms
-          .filter((mTerm) => mTerm.type !== syntheticLabel)
-          .map((mTerm) => {
-            if (saveOneAlignment(mTerm)) {
-              savedTerms++;
-            }
-          })
+        alignments.map((mTerm) => {
+          if (saveOneAlignment(mTerm)) {
```
Comment (excelsior) [inline app/javascript/components/align-and-fine-tune/AlignAndFineTune.jsx:506]: `saveOneAlignment` is an async function, without `await` it'll always return a truthy value here.

**Review by excelsior — CHANGES_REQUESTED** [review]
Thanks @richard0000 

Recent changes (comments etc) work fine. There's still a problem with error handling, though. Also, it seems that a redirect now happens when at least one term is successfully updated. Wouldn't it be better the other way around—redirect the user only when there are no errors at all and show the alert otherwise?

**Comment by richard0000** [comment]
> Thanks @richard0000
> 
> Recent changes (comments etc) work fine. There's still a problem with error handling, though. Also, it seems that a redirect now happens when at least one term is successfully updated. Wouldn't it be better the other way around—redirect the user only when there are no errors at all and show the alert otherwise?

@excelsior look at lines 460 and 467. Errors will interrupt the operation and will not redirect the user.

File: app/controllers/api/v1/mapping_terms_controller.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  ###
+  def update
+    ActiveRecord::Base.transaction do
+      if params[:mapping_term][:mapped_terms].present?
+        @mapping_term.update_mapped_terms(params[:mapping_term][:mapped_terms])
+      end
+      @mapping_term.update!(permitted_params)
+    end
+
+    render json: @mapping_term, inlcude: %i[mapped_terms]
```
Comment (excelsior) [inline app/controllers/api/v1/mapping_terms_controller.rb:20]: There's a typo: `include`.

**Review by excelsior — APPROVED** [review]

**Pull request merged**